Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Lib/LocationUpdates.hs

```diff
@@ -55,66 +71,84 @@ checkForDeviation routeDeviationThreshold estimatedRoute (pt : batchWaypoints) d
       then checkForDeviation routeDeviationThreshold estimatedRoute batchWaypoints 0
       else checkForDeviation routeDeviationThreshold estimatedRoute batchWaypoints (deviationCount + 1)
 
-updateDeviation :: (HedisFlow m r, CacheFlow m r, EsqDBFlow m r, EsqDBReplicaFlow m r, EncFlow m r) => Meters -> Maybe (Id Ride) -> [LatLong] -> m Bool
-updateDeviation _ Nothing _ = do
+updateDeviation :: TransporterConfig -> Bool -> Maybe (Id Ride) -> [LatLong] -> Flow Bool
+updateDeviation _ _ Nothing _ = do
   logInfo "No ride found to check deviation"
   return False
-updateDeviation routeDeviationThreshold (Just rideId) batchWaypoints = do
-  routeDeviation <- getRouteDeviation rideId
-  if routeDeviation
+updateDeviation transportConfig safetyCheckEnabled (Just rideId) batchWaypoints = do
+  (routeDeviation, safetyAlertTriggered) <- getDeviationAndSafetyDetails rideId
+  if safetyAlertTriggered
     then do
-      logInfo $ "Ride already deviated for rideId: " <> getId rideId
+      logInfo $ "Safety alert already triggered for rideId: " <> getId rideId
       return True
     else do
-      logInfo $ "Updating Deviation for rideId: " <> getId rideId
-      let key = searchRequestKey (getId rideId)
+      let routeDeviationThreshold = transportConfig.routeDeviationThreshold
+          key = searchRequestKey (getId rideId)
       routeInfo :: Maybe RI.RouteInfo <- Redis.get key
       case routeInfo >>= (.points) of
-        Just estimatedRoute ->
-          if checkForDeviation routeDeviationThreshold estimatedRoute batchWaypoints 0
-            then do
-              logInfo $ "Deviation detected for rideId: " <> getId rideId
-              QRide.updateDriverDeviatedFromRoute rideId True
-              return True
+        Just estimatedRoute -> do
+          isDeviation <-
+            if safetyCheckEnabled
+              then do
+                let dummyHighDeviationValue = transportConfig.safetyRouteDeviationThreshold * 2
+                    deviation = getMaxDeviation estimatedRoute batchWaypoints dummyHighDeviationValue dummyHighDeviationValue 0
+                fork "Performing safety check" $
+                  when (deviation >= transportConfig.safetyRouteDeviationThreshold) $ performSafetyCheck transportConfig rideId
+                return $ deviation >= routeDeviationThreshold
```

Comment:
Should we also update driverDeviatedFromRoute in ride table if we are returning true?